### PR TITLE
Async Workflows  in Agency Swarm

### PR DESCRIPTION
**Pull request by gregnwosu**
first cut of async tools and agents in agency swarm. 

A couple of things to bear in mind
- the async agency class needs to have initialize called on it before use. this is because init_oai needs to be done asynchronously and hence cant be done from the constructor.
- i have used pytest for a couple of basic tests of the async agency, the tests are not comprehensive and could do with extendeding
- i havent fully tested the asymc tools, i have created one called useless tool, in the pytest fixtures.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -2,4 +2,7 @@ openai==1.3.0
 instructor==0.3.4
 deepdiff==6.7.1
 termcolor==2.3.0
-python-dotenv==1.0.0
\ No newline at end of file
+python-dotenv==1.0.0
+pytest~=7.4.3
```
Comment (nicko-ai) [inline requirements.txt:6]: pytest is a dev dependency. Should be separated. I suggest using poetry for package and dependency management

File: tests/agency/conftest.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+def agency(agency_chart, shared_instructions):
+    return Agency(agency_chart, shared_instructions)
+
+
+@pytest.fixture
+def async_agency(async_agency_chart, shared_instructions):
+    agency = AsyncAgency(async_agency_chart, shared_instructions)
+    return agency
+
+
```
Comment (nicko-ai) [inline tests/agency/conftest.py:94]: need to apply ruff or similar tool to fix such problems.

Here's the example config:
.pre-commit-config.yaml:
```
repos:
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.5.0
    hooks:
      - id: trailing-whitespace
      - id: end-of-file-fixer
      - id: check-yaml
      - id: check-toml
      - id: debug-statements
        language_version: python3

  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.1.6
    hooks:
      - id: ruff
        args: [--fix]
      - id: ruff-format

  - repo: https://github.com/pre-commit/mirrors-mypy
    rev: v1.7.0
    hooks:
      - id: mypy
        args:
          - --no-warn-no-return
          - --ignore-missing-imports
```

`pip install pre-commit`, then `pre-commit install-hooks`, trigger manually: `pre-commit run`

**Comment by VRSEN** [comment]
Thanks guys, I will review shortly

File: tests/agency/conftest.py (hunk 2 above)
Comment (gregnwosu) [reply tests/agency/conftest.py:94]: Hi I know its not perfect, could you possibly help by sending me the config that this needs or updating the branch?

File: requirements.txt (hunk 1 above)
Comment (gregnwosu) [reply requirements.txt:6]: Hi I know its not perfect, could you possibly help by sending me the config that this needs or updating the branch?

File: tests/agency/conftest.py (hunk 2 above)
Comment (nicko-ai) [reply tests/agency/conftest.py:94]: just run the commands above to fix the code. The config is above, you don't need anything else.
In addition, your IDE should automatically pick it up, e.g. PyCharm.

Comment (gregnwosu) [reply tests/agency/conftest.py:94]: will do!

**Comment by gregnwosu** [comment]
i think i need to understand the framework more

**Pull request closed without merging**